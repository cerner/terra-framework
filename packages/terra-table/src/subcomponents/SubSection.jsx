// TODO: fix linter error
/* eslint-disable jsx-a11y/control-has-associated-label */

import React, { useCallback, useContext } from 'react';
import classNames from 'classnames/bind';
import PropTypes from 'prop-types';

import ThemeContext from 'terra-theme-context';
import GridContext, { GridConstants } from '../utils/GridContext';

import columnShape from '../proptypes/columnShape';
import styles from './SubSection.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * An identifier for the section. This identifier should be unique across the set of sections provided to the DataGrid.
   */
  id: PropTypes.string.isRequired,

  /**
   * Unique identifier for the parent table
   */
  tableId: PropTypes.string.isRequired,

  /**
   * The subsection's row index position in the table.
   */
   subSectionRowIndex: PropTypes.number,

  /**
   * A text string to render within the section header.
   */
  text: PropTypes.string,

  /**
   * All columns currently displayed.
   */
  displayedColumns: PropTypes.arrayOf(columnShape),
};

function Section(props) {
  const {
    id,
    tableId,
    subSectionRowIndex,
    text,
    displayedColumns,
  } = props;

  const theme = useContext(ThemeContext);

  const gridContext = useContext(GridContext);
  const isGridContext = gridContext.role === GridConstants.GRID;

  return (
    <>
      <tr
        aria-rowindex={subSectionRowIndex}
        className={cx('header-row', theme.className)}
        data-section-id={id}
      >
        <th
          id={`${tableId}-${id}`}
          className={cx('header-cell')}
          align="left"
          colSpan={displayedColumns.length}
          role="columnheader"
          scope="col"
          tabIndex={isGridContext ? -1 : undefined}
        >
          <h3
            className={cx('subsection')}
          >
            <span
              className={cx('sticky')}
            >
              {text}
            </span>
          </h3>
        </th>
      </tr>
    </>
  );
}

Section.propTypes = propTypes;

export default React.memo(Section);
