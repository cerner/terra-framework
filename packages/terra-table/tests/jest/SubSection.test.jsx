import React from 'react';
import SubSection from '../../src/subcomponents/SubSection';

const cols = [
  { id: 'Column-0', displayName: 'Vitals' },
  { id: 'Column-1', displayName: 'March 16' },
  { id: 'Column-2', displayName: 'March 17', isSelectable: false },
];

describe('SubSection', () => {
  it('verifies that the table created is consistent with the rows and overflowColumns props', () => {
    const wrapper = enzyme.shallow(
      <SubSection
        id="subsection-0"
        tableId="table-0"
        subSectionRowIndex={0}
        text="Test SubSection"
        displayedColumns={cols}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
