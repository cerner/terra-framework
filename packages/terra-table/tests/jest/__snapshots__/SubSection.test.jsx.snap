// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubSection verifies that the table created is consistent with the rows and overflowColumns props 1`] = `
<Fragment>
  <tr
    aria-rowindex={0}
    className="header-row"
    data-section-id="subsection-0"
  >
    <th
      align="left"
      className="header-cell"
      colSpan={3}
      id="table-0-subsection-0"
      role="columnheader"
      scope="col"
    >
      <h3
        className="subsection"
      >
        <span
          className="sticky"
        >
          Test SubSection
        </span>
      </h3>
    </th>
  </tr>
</Fragment>
`;
